SI-7715 Etapolator s"$_"

som-snytt opened the pull request:
Allows `vs map s"How many? $_"`.

Or `vs zip vs map s"How many? $_ by $_".tupled`.

And `s"How many? $_ by $_ by $_"("one","two","three")`.

In a pattern,

```
    scala> val rx"$_ by ${I(b)}" = "2 by 4"
    b: Int = 4
```

scala/bug#7715

som-snytt committed:
SI-7715 Etapolator s"$_"

Allows `vs map s"How many? $_"`.

Or `vs zip vs map s"How many? $_ by $_".tupled`.

And `s"How many? $_ by $_ by $_"("one","two","three")`.

In a pattern,

```
scala> val rx"$_ by ${I(b)}" = "2 by 4"
b: Int = 4

```

xeno-by commented:
Man, I'm amazed at your word synthesis imagination.

xeno-by commented on test/files/run/interpolation.flags:
```diff
@@ -1 +0,0 @@
--Xexperimental
```
I'd move this into a separate commit

xeno-by commented on test/files/run/t7715-etapolator.scala:
```diff
@@ -0,0 +1,39 @@
+    case _                      => -1
+  })
+  Console println ("2 by 4" match {
+    case rx"$_ by ${I(b)}"      => b    // is emitted this way, too
```
Oh man, isn't this wonderful?!

som-snytt commented:
Recently, people were asking for language support for updating vars with tuple pattern assignment syntax. I just wrote my first macro with quasiquotes to do that.  I call it: The Intupolator!  Anyway, quasiquotes are pretty neat.

```
var a, b, c = 0
def pairing: Product = (11, 12)
u(a, b) = (7, 8)
u(a, b) = pairing
//u(a, b, c) = pairing //java.lang.AssertionError: assertion failed: Bad product arity for assignment
```

som-snytt committed:
StringContext#checkLengths reports bad args count

Since interpolator args are of type Any, it's easy
to write s(args) instead of s(args: _*).  I wonder
if Xlint would have warned me about that.

som-snytt committed:
In tests, interpolation is no longer -Xperimental.

retronym commented:
This seems irregular:

```
% qbin/scala
Welcome to Scala version 2.11.0-20130803-110323-60fa47c923 (Java HotSpot(TM) 64-Bit Server VM, Java 1.6.0_37).
scala> val template = s"""This is string $_ """
template: Any => String = <function1>

scala> val template = s"""This is string ${_} """
<console>:1: error: unbound placeholder parameter
       val template = s"""This is string ${_} """
                                           ^
```

Furthermore, it would be good to allow type ascription of the holes.

```
scala> val template = s"""This is string ${_: Int} """
<console>:1: error: unbound placeholder parameter
       val template = s"""This is string ${_: Int} """
                                           ^
```

Here's the original [discussion of this feature on [scala-sips]](https://groups.google.com/d/msg/scala-sips/DefqJzh0ux0/E7sKJYgWBHsJ)

som-snytt commented:
Yes, I observed the first discrepancy, but this change was only for $_.  The Any is glued on early.  But I guess it can happen after an { expr0 } hole.

Thanks for the link, I missed that.  Figures this was already on the long-range sensors.  I had just noticed that $_foo doesn't interpolate, so I thought $_ should do the obvious.  I'm always the last to the party.  Two years late, in this case (dating it from Miles' gist). I also hadn't noticed 6010 for some reason.

retronym commented:
I'm a little bit hesitant allowing this in non-pattern contexts. The placeholder syntax for anonymous functions is a common source of confusion ("Which enclosing expression binds this `_`?"). Adding the string interpolation desugaring into the mix seems like a way to keep the style guide authors busy penning new "don't do this" entries.

So we should convince ourselves that `s"$_"  is worth it as compared to`(a: Int) => s"$a"`. I'll put the question to the Scala meeting tomorrow.

som-snytt commented:
The double quotes are pretty encapsulating.

Here's where I needed it recently:

https://github.com/scala/scala/commit/5edb79827779dbd983f9fac37e8fd926afdaba04#L0R76

Or as King Lear had it, Reason not the need.

som-snytt committed:
SI-7544 Interpolation message for %% literal

The new error text: "conversions must follow a splice; use %% for
literal %, %n for newline".

The error is emitted when % is not at the start of a part; only
the non-conversions %% and %n are allowed.

It would be nice if it were convenient to parse the part to see
what the user may have intended, but c'est l'erreur.

som-snytt committed:
SI-7715 Interpolation allows ${_} and ${_: Int}

The placeholders in interpolation args are local to the
interpolation.

Without an explicit type, Any is implied by definition.

```
scala> s"who ${_}"
res0: Any => String = <function1>

scala> s"who ${_ : Int}"
res1: Int => String = <function1>

scala> res1(4)
res2: String = who 4

scala> s"who ${_ : Int}  ${_: Double}"
res3: (Int, Double) => String = <function2>

scala> res3(4,5.0)
res4: String = who 4  5.0

scala> res3(4,5)
res5: String = who 4  5.0

scala> res3(4,"foo")
<console>:9: error: type mismatch;
 found   : String("foo")
 required: Double
              res3(4,"foo")
                     ^

scala> s"${ if (_: Boolean) _ else _ }"
res7: (Boolean, Any, Any) => String = <function3>

scala> res7(true, "foo", "bar")
res8: String = foo

```

som-snytt commented:
Besides, things are getting a bit slow over at scalapuzzlers.com.

But seriously, this seems pretty innocent at first glance, compared to type inference + implicits mind benders like a really naive one I squandered time on yesterday at SO.  (By squander, I mean I tried to figger it out but was too dumb, so I'm still unenlightened.) (Also by squander, I mean I missed Double Life of Veronique on Hulu.)

adriaanm commented:
I agree with @retronym. The underscore makes a lot of sense in pattern matches, but I don't think we should further complicate the creation of function literals. It's already way too complicated as it is.

adriaanm commented:
I'm going to close this PR, but I (and uninspired pattern namers everywhere) welcome one limited to pattern scope.

I hereby propose the following rule: "The underscore has no meaning -- it can be replaced by a fresh identifier without changing the semantics of your program."

The only exception I think is worthwhile is in turning partially applied methods into functions.

som-snytt commented:
@adriaanm @retronym I just encountered another use case, an out-of-line splice that looks better this way, without extra `(someVar: String, someOtherVar: String) =>` in front, and certainly better than in-line:

```
      val preamble = s"""
        |$headerPreamble
        |$_
        |object ${lineRep.readName} $_ {
        |${envLines mkString ("  ", "\n  ", "\n")}
        |$importsPreamble
        |${indentCode(toCompute)}""" (
          if (importsSuper.nonEmpty) s"trait ${lineRep.readName}Implicits { $importsSuper }" else "",
          if (importsSuper.nonEmpty) s"extends ${lineRep.readName}Implicits " else "",
        ).stripMargin
```

Maybe that wouldn't be the final form, but it's easy to see how much easier it is to read the long exprs out-of-line, even (or especially) with the placeholders holding their own amongst many other splicées (excuse my French).

Without the function $_yntax, you're forced to create extra tmp variables, either params or locals.

gkossakowski commented:
@som-snytt: I think your snippet shows very well how `$_` syntax can be abused. You really jammed way too much in a single expression. The fact that you have to create a temporary variables would help with readability of your snippet in my opinion.

som-snytt commented:
@gkossakowski I see what you mean, but I think it shows the template has a certain shape, with a little bit that is a function of something.  (And the little bit may be empty.)  Of course, really the whole template is a function of all the args to `s`.

The snippet doesn't actually compile because I can't invoke String.nonEmpty.

```
[quick.repl]  both method stringToTermName in trait Names of type (s: String)IMain.this.global.TermName
[quick.repl]  and method stringToTypeName in trait Names of type (s: String)IMain.this.global.TypeName
[quick.repl]  are possible conversion functions from Request.this.importsSuper.type to ?{def nonEmpty: ?}
```

I guess there are degrees of complexity.

som-snytt commented:
s"$_" is something that is too much fun not to exist.

som-snytt commented:
Another comparison.

```
  val responses = List(english -> ((x: String) => s"goodbye, $x"), deutsch -> ((x: String) => s"auf wiedersehen, $x"))
  //val responses = List(english -> s"goodbye, $_", deutsch -> s"auf wiedersehen, $_")
```

erikkaplun commented:
I know I'm nobody here... but it would be very nice to see this re-opened and re-considered, as this is something that would make the language more consistent.

adriaanm commented:
Hi @eallik! Thanks for your feedback. Agree this would be somewhat more regular, but that has to be balanced with how easy it is to understand and explain (what should we infer for the types of the arguments?). Btw, the pattern matching part of this was merged in #2823.
